crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -1089,41 +1093,39 @@ where
                 cases,
                 range: _,
             }) => {
+                debug_assert!(self.current_match_case.is_none());
+
                 let subject_expr = self.add_standalone_expression(subject);
                 self.visit_expr(subject);
-
-                let after_subject = self.flow_snapshot();
-                let Some((first, remaining)) = cases.split_first() else {
+                if cases.is_empty() {
                     return;
                 };
 
-                let first_constraint_id = self.add_pattern_constraint(
-                    subject_expr,
-                    &first.pattern,
-                    first.guard.as_deref(),
-                );
-
-                self.visit_match_case(first);
-
-                let first_vis_constraint_id =
-                    self.record_visibility_constraint(first_constraint_id);
-                let mut vis_constraints = vec![first_vis_constraint_id];
-
+                let after_subject = self.flow_snapshot();
+                let mut vis_constraints = vec![];
                 let mut post_case_snapshots = vec![];
-                for case in remaining {
-                    post_case_snapshots.push(self.flow_snapshot());
-                    self.flow_restore(after_subject.clone());
+                for (i, case) in cases.iter().enumerate() {
+                    if i != 0 {
+                        post_case_snapshots.push(self.flow_snapshot());
+                        self.flow_restore(after_subject.clone());
+                    }
+
+                    self.current_match_case = Some(CurrentMatchCase::new(&case.pattern));
+                    self.visit_pattern(&case.pattern);
+                    self.current_match_case = None;
                     let constraint_id = self.add_pattern_constraint(
                         subject_expr,
                         &case.pattern,
                         case.guard.as_deref(),
                     );
-                    self.visit_match_case(case);
-
                     for id in &vis_constraints {
                         self.record_negated_visibility_constraint(*id);
                     }
                     let vis_constraint_id = self.record_visibility_constraint(constraint_id);
+                    if let Some(expr) = &case.guard {
+                        self.visit_expr(expr);
+                    }
+                    self.visit_body(&case.body);
```

Comment:
h/t to @sharkdp for noticing that we were performing these steps out of order.  The correct order is:

1. visit the pattern first
2. add the pattern constraint
3. visit the guard expression and body

That is, each pattern's constrain applies to the guard and body of that match case, but not to the pattern itself.

Also note that this is different than what David first suggested, which was that the pattern constraint should not apply to the guard either.  I think that it _should_ apply to the guard, giving my reading of this part of [the spec](https://docs.python.org/3/reference/compound_stmts.html#overview):

> If the pattern succeeds, the corresponding guard (if present) is evaluated. In this case all name bindings are guaranteed to have happened.

Reply:
> Also note that this is different than what David first suggested, which was that the pattern constraint should not apply to the guard either. I think that it should apply to the guard

Ah, yes. Of course.